consensus/hotstuff/cruisectl/block_rate_controller.go

```diff
@@ -0,0 +1,160 @@
+// Package cruisectl implements a "cruise control" system for Flow by adjusting
+// nodes' block rate delay in response to changes in the measured block rate.
+//
+// It uses a PID controller with the block rate as the process variable and
+// the set-point computed using the current view and epoch length config.
+package cruisectl
+
+import (
+	"time"
+
+	"github.com/rs/zerolog"
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/state/protocol"
+)
+
+// measurement represents one measurement of block rate and error.
+// A measurement is taken each time the view changes for any reason.
+// Each measurement measures the instantaneous and exponentially weighted
+// moving average (EWMA) block rates, computes the target block rate,
+// and computes the error terms.
+type measurement struct {
+	view            uint64    // v       - the current view
+	time            time.Time // t[v]    - when we entered view v
+	blockRate       float64   // r[v]    - measured instantaneous block rate at view v
+	aveBlockRate    float64   // r_N[v]  - EWMA block rate over past views [v-N, v]
+	targetBlockRate float64   // r_SP[v] - computed target block rate at view v
+	proportionalErr float64   // e_N[v]  - proportional error at view v
+	integralErr     float64   // E_N[v]  - integral of error at view v
+	derivativeErr   float64   // ∆_N[v]  - derivative of error at view v
+}
+
+// epochInfo stores data about the current and next epoch. It is updated when we enter
+// the first view of a new epoch, or the EpochSetup phase of the current epoch.
+type epochInfo struct {
+	curEpochFinalView        uint64
+	curEpochTargetSwitchover time.Time
+	nextEpochFinalView       *uint64
+	epochFallbackTriggered   *atomic.Bool
+}
+
+// BlockRateController dynamically adjusts the block rate delay of this node,
+// based on the measured block rate of the consensus committee as a whole, in
+// order to achieve a target overall block rate.
+type BlockRateController struct {
+	component.Component
+
+	config *Config
+	state  protocol.State
+	log    zerolog.Logger
+
+	lastMeasurement *measurement    // the most recently taken measurement
+	blockRateDelay  *atomic.Float64 // the block rate delay value to use when proposing a block
```

Comment:
same here, why pointer reference?

Reply:
I was under the impression that atomic variables needed to be protected manually from copying, which storing them as pointers as a general rule accomplishes. We do this frequently in our codebase (for example, the existing `BlockRateDelay` field is an `*atomic.Float64` type. But there is actually a `go vet` rule which checks for this by looking for a special `nocopy` type. If an atomic is copied, the code won't compile. (Only for stdlib atomics though, Uber's library doesn't have access to the same protection.)

Reply:
I have re-arranged some of these variables in the follow-up PR https://github.com/onflow/flow-go/pull/4348. I will adjust the pointer/value types as necessary there.